apstools/filewriters.py

```diff
@@ -689,3 +703,1011 @@ def myPlan():
+    =========== ============= ===================================================
+    
+    Notes:
+    
+    1. ``detectors[0]`` will be used as the ``/entry/data@signal`` attribute
+    2. the *complete* list in ``positioners`` will be used as the ``/entry/data@axes`` attribute
+
+    METHODS
+
+    .. autosummary::
+       
+       ~writer
+       ~h5string
+       ~add_dataset_attributes
+       ~assign_signal_type
+       ~create_NX_group
+       ~get_sample_title
+       ~get_stream_link
+       ~write_data
+       ~write_detector
+       ~write_entry
+       ~write_instrument
+       ~write_metadata
+       ~write_monochromator
+       ~write_positioner
+       ~write_root
+       ~write_sample
+       ~write_slits
+       ~write_source
+       ~write_streams
+       ~write_user
+    """
+    file_extension = NEXUS_FILE_EXTENSION
+    instrument_name = None      # name of this instrument
+    nexus_release = NEXUS_RELEASE  # # NeXus release to which this file is written
+    nxdata_signal = None        # name of dataset for Y axis on plot
+    nxdata_signal_axes = None   # name of dataset for X axis on plot
+    root = None                 # instance of h5py.File
+
+    # convention: methods written in alphabetical order
+
+    def add_dataset_attributes(self, ds, v, long_name=None):
+        """
+        add attributes from v dictionary to dataset ds
+        """
+        ds.attrs["units"] = self.h5string(v["units"])
+        ds.attrs["source"] = self.h5string(v["source"])
+        if long_name is not None:
+            ds.attrs["long_name"] = self.h5string(long_name)
+        if v["dtype"] not in ("string",):
+            ds.attrs["precision"] = v["precision"]
+            def cautious_set(key):
+                if v[key] is not None:
+                    ds.attrs[key] = v[key]
+            cautious_set("lower_ctrl_limit")
+            cautious_set("upper_ctrl_limit")
+ 
+    def assign_signal_type(self):
+        """
+        decide if a signal in the primary stream is a detector or a positioner
+        """
+        try:
+            primary = self.root["/entry/instrument/bluesky/streams/primary"]
+        except KeyError:
+            raise KeyError(
+                f"no primary data stream in "
+                f"scan {self.scan_id} ({self.uid[:7]})"
+                )
+        for k, v in primary.items():
+            # logger.debug(v.name)
+            # logger.debug(v.keys())
+            if k in self.detectors:
+                signal_type = "detector"
+            elif k in self.positioners:
+                signal_type = "positioner"
+            else:
+                signal_type = "other"
+            v.attrs["signal_type"] = signal_type            # group
+            try:
+                v["value"].attrs["signal_type"] = signal_type   # dataset
+            except KeyError:
+                logger.warning("Could not assign %s as signal type %s", k, signal_type)
+
+    def create_NX_group(self, parent, specification):
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/gh/BCDA-APS/apstools/pullRequest?prid=5802262)

Reply:
Will leave as a method.  Ignore Codacy report.